hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/BeanValidationEventListener.java

```diff
@@ -110,11 +114,21 @@ public void onPreUpdateCollection(PreCollectionUpdateEvent event) {
 		final Object entity = castNonNull( event.getCollection().getOwner() );
 		validate(
 				entity,
-				event.getSession().getEntityPersister( event.getAffectedOwnerEntityName(), entity ),
+				getEntityPersister( event.getSession(), event.getAffectedOwnerEntityName(), entity ),
 				GroupsPerOperation.Operation.UPDATE
 		);
 	}
 
+	private EntityPersister getEntityPersister(SharedSessionContractImplementor session, String entityName, Object entity) {
+		if ( session != null ) {
+			return session.getEntityPersister( entityName, entity );
+		}
+		return entityName == null
+				? sessionFactory.getMappingMetamodel().getEntityDescriptor( entity.getClass().getName() )
+				: sessionFactory.getMappingMetamodel().getEntityDescriptor( entityName )
+						.getSubclassEntityPersister( entity, sessionFactory );
+	}
```

Comment:
This is a fine workaround, but I wonder if we could rather make `org.hibernate.event.spi.AbstractSessionEvent` take a `org.hibernate.engine.spi.SharedSessionContractImplementor` instead of `org.hibernate.event.spi.EventSource`, deprecating the old signature (and delegating to the new one since `EventSource` extends `SharedSessionContractImplementor`.

Since stateless session fire this events too now, it would make more sense to be able to retrieve them in `AbstractSessionEvent#getSession` (which, as the name implies, could not only return `EventSource`s). Perhaps let's hear what @gavinking thinks.

Reply:
+1 there is the 

https://github.com/hibernate/hibernate-orm/blob/1753cb6c6712c7a0525a2905f064a5ae73076743/hibernate-core/src/main/java/org/hibernate/event/spi/AbstractEvent.java#L18-L19

but it's a different hierarchy of events, and these collection ones don't get the same source 😖